orchestrator/services/processes.py

```diff
@@ -155,54 +138,101 @@ def _db_log_step(
             else:
                 p.assignee = Assignee.SYSTEM
 
-        # check if last error state is identical to determine if we add a new step or update the last one
-        last_db_step = p.steps[-1] if len(p.steps) else None
+    else:
+        p.failed_reason = None
+        p.traceback = None
+
+    return p
+
 
+def _get_current_step_to_update(
+    stat: ProcessStat, p: ProcessTable, step: Step, process_state: WFProcess
+) -> ProcessStepTable:
+    """Checks if last error state is identical to determine if we add a new step or update the last one."""
+    step_state: State = process_state.unwrap()
+    current_step = None
+    last_db_step = p.steps[-1] if len(p.steps) else None
+    current_sub_step, current_step_group = step_state.get("__sub_step"), step_state.get("__step_group")
+    db_sub_step, db_step_group = (
+        (last_db_step.state.get("__sub_step"), last_db_step.state.get("__step_group")) if last_db_step else (None, None)
+    )
+
+    if current_sub_step is None and db_sub_step is None:
+        pass  # Skip step group logic
+    elif (db_sub_step is None and current_sub_step) or (db_sub_step and current_sub_step is None):
```

Comment:
I wonder what are the possible scenarios to reach this code.

Does `(db_sub_step is None and current_sub_step)` mean we've just entered a sub-step in the step-group and nothing was yet logged to the database?

And then does `(db_sub_step and current_sub_step is None)` mean we've resumed/continued a sub-step that was already in the database?

Reply:
Yes and no. The `current_sub_step` is the sub step as recorded in the state after the last step. If this is not present, it means we are not currently in a sub step. If `db_sub_step` is present, it means we just finished the last sub step of the step group.

```mermaid
stateDiagram-v2

    [*] --> exec_steps
    exec_steps --> a
    exec_steps --> b
    exec_steps --> f
    b --> exec_steps(2)
   exec_steps(2) --> c
   exec_steps(2) --> d
   exec_steps(2) --> e
```

- `(db_sub_step is None and current_sub_step)` happens when we just executed `c`
- `(db_sub_step and current_sub_step is None)` happens when we just executed `e`

I've been thinking on how to refactor this, but one change could be to just automatically add a sub_step to the start and end of the sub_steps chain to add/remove sub_step related data to the state. Then this logic could be simplified. The db_logstep does way too much at the moment.